src/Node/Expression/NameExpression.php

```diff
@@ -52,7 +52,7 @@ public function compile(Compiler $compiler): void
                 ;
             }
         } elseif ($this->isSpecial()) {
-            $compiler->raw($this->specialVars[$name]);
+            $compiler->raw(static::$specialVars[$name]);
```

Comment:
accessing a private property with `static` is invalid. It would break if the class is extended.

Reply:
My friend building a SAAS platform, and searching for something similar to [Liquid](https://github.com/Shopify/liquid) but for PHP. I suggested `Twig` to him, but he found that special vars cannot be disabled. These variables expose some data that shouldn't be shown for his clients.

In short, the use case is stated in the first point of liquid's pros:
> - You want to allow your users to edit the appearance of your application but don't want them to run insecure code on your server.
... or see private data like (The full path of view, the context object) 

Using static variables is a way followed by some other packages to allow customization, like [Tenancy For Laravel](https://tenancyforlaravel.com/docs/v3/configuration/) package

> ... whenever the class you're using has a public static property, it's intended to be configured.

Reply:
Which data is it that should not be shown ?
Blocking access to `_context` won't prevent access to the content of the context (they can try to find the available variables instead of looping over this special variable to discover them all). The Twig context is just the associative array of variables available in the template. If they contain things that users should not see, you should not inject them in the template in the first place as they can access the normal variable anyway.
The `_charset` variable is not really sensitive (in 99% of cases, this will give the string `UTF-8`).
`_self` return the template **name**, not the full path (you could of course register a FilesystemLoader with its root directory being the root of the filesystem, in which case the name would indeed be the full path, but this is then an issue on your side as `source('...')` would then allow reading any file. That's not the recommended setup, especially not if you have to deal with untrusted templates).

> Using static variables is a way followed by some other packages to allow customization

I see 2 issues here:
- static variables are a global state, so this is a global configuration, which would leak to other usages of Twig (you cannot have 2 Twig Environment with different static variables)
- your quote talks about *public* properties. Private static properties cannot be configured.